test/testdriver.go

```diff
@@ -14,32 +15,49 @@ type SoftwarePainter interface {
 }
 
 type testDriver struct {
-	device  *device
-	painter SoftwarePainter
+	device       *device
+	painter      SoftwarePainter
+	windows      []fyne.Window
+	windowsMutex sync.RWMutex
 }
 
 // Declare conformity with Driver
 var _ fyne.Driver = (*testDriver)(nil)
 
 func (d *testDriver) CanvasForObject(fyne.CanvasObject) fyne.Canvas {
-	windowsMutex.RLock()
-	defer windowsMutex.RUnlock()
+	d.windowsMutex.RLock()
+	defer d.windowsMutex.RUnlock()
 	// cheating as we only have a single test window
-	return windows[0].Canvas()
+	return d.windows[0].Canvas()
 }
 
 func (d *testDriver) AbsolutePositionForObject(co fyne.CanvasObject) fyne.Position {
 	return co.Position() // TODO get the real answer
 }
 
 func (d *testDriver) CreateWindow(string) fyne.Window {
-	return NewWindow(nil)
+	canvas := NewCanvas().(*testCanvas)
+	if fyne.CurrentApp() != nil {
```

Comment:
these 2 lines are really complex - why not just use the "d" *testDriver?

Reply:
I see its because of the refactor - but this is an opportunity to remove the complexity :)